hal/aurix_tc3xx.c

```diff
@@ -0,0 +1,511 @@
+#include "image.h"  /* for RAMFUNCTION */
+#include "loader.h" /* for wolfBoot_panic */
+
+/* ILLD headers */
+#include "IfxCpu_reg.h"    /* for CPU0_FLASHCON1 */
+#include "IfxFlash.h"      /* for IfxFlash_eraseMultipleSectors, */
+#include "IfxPort.h"       /* for IfxPort_*  */
+#include "IfxScuRcu.h"     /* for IfxScuRcu_performReset */
+#include "Ifx_Ssw_Infra.h" /* for Ifx_Ssw_jumpToFunction */
+
+#define FLASH_MODULE                (0)
+#define UNUSED_PARAMETER            (0)
+#define WOLFBOOT_AURIX_RESET_REASON (0x5742) /* "WB" */
+
+/* Helper macros to gets the base address of the page, wordline, or sector that
+ * contains byteAddress */
+#define GET_PAGE_ADDR(addr) \
+    ((uintptr_t)(addr) & ~(IFXFLASH_PFLASH_PAGE_LENGTH - 1))
+#define GET_WORDLINE_ADDR(addr) \
+    ((uintptr_t)(addr) & ~(IFXFLASH_PFLASH_WORDLINE_LENGTH - 1))
+#define GET_SECTOR_ADDR(addr) ((uintptr_t)(addr) & ~(WOLFBOOT_SECTOR_SIZE - 1))
+
+/* RAM buffer to hold the contents of an entire flash sector*/
+static uint32_t sectorBuffer[WOLFBOOT_SECTOR_SIZE / sizeof(uint32_t)];
+
+#define LED_PROG     (0)
+#define LED_ERASE    (1)
+#define LED_READ     (2)
+#define LED_WOLFBOOT (5)
+
+#ifdef WOLFBOOT_AURIX_GPIO_TIMING
+#ifndef SWAP_LED_POLARITY
+#define LED_ON(led)  IfxPort_setPinLow(&MODULE_P00, (led))
+#define LED_OFF(led) IfxPort_setPinHigh(&MODULE_P00, (led))
+#else
+#define LED_ON(led)  IfxPort_setPinHigh(&MODULE_P00, (led))
+#define LED_OFF(led) IfxPort_setPinLow(&MODULE_P00, (led))
+#endif
+#else
+#define LED_ON(led)
+#define LED_OFF(led)
+#endif /* WOLFBOOT_AURIX_GPIO_TIMING */
+
+/* Returns the SDK flash type enum based on the address */
+static IfxFlash_FlashType getFlashTypeFromAddr(uint32_t addr)
+{
+    IfxFlash_FlashType type = 0;
+
+    if (addr >= IFXFLASH_DFLASH_START && addr <= IFXFLASH_DFLASH_END) {
+        /* Assuming D0 for simplicity */
+        type = IfxFlash_FlashType_D0;
+    }
+    else if (addr >= IFXFLASH_PFLASH_P0_START && addr <= IFXFLASH_PFLASH_P0_END)
+    {
+        type = IfxFlash_FlashType_P0;
+    }
+    else if (addr >= IFXFLASH_PFLASH_P1_START && addr <= IFXFLASH_PFLASH_P1_END)
+    {
+        type = IfxFlash_FlashType_P1;
+    }
+    else {
+        /* bad address, panic for now */
+        wolfBoot_panic();
+    }
+
+    return type;
+}
+
+/* Programs a single page in flash */
+static void RAMFUNCTION programPage(uint32_t address,
+                                    const uint32_t* data,
+                                    IfxFlash_FlashType type)
+{
+    if (address % IFXFLASH_PFLASH_PAGE_LENGTH != 0) {
+        wolfBoot_panic();
+    }
+
+    const uint16 endInitSafetyPassword =
+        IfxScuWdt_getSafetyWatchdogPasswordInline();
+
+    IfxFlash_enterPageMode(address);
+    IfxFlash_waitUnbusy(FLASH_MODULE, type);
+
+    for (size_t offset = 0;
+         offset < IFXFLASH_PFLASH_PAGE_LENGTH / sizeof(uint32_t);
+         offset += 2)
+    {
+        IfxFlash_loadPage2X32(address, data[offset], data[offset + 1]);
```

Comment:
size_t offset. Mixed declaration. Move variable definitions to top.

Reply:
done